src/crypto/tls/delegated_credentials.go

```diff
@@ -0,0 +1,489 @@
+// this can be used to delegate TLS termination to a reverse proxy. Credentials
+// can't be revoked; in order to mitigate risk in case the middlebox is
+// compromised, the credential is only valid for a short time (days, hours, or
+// even minutes).
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"time"
+)
+
+const (
+	dcMaxTTLSeconds   = 60 * 60 * 24 * 7 // The maxium validity period is 7 days
+	dcMaxTTL          = time.Duration(dcMaxTTLSeconds * time.Second)
+	dcMaxPubLen       = 1 << 24 // Bytes
+	dcMaxSignatureLen = 1 << 16 // Bytes
+)
+
+var errNoDelegationUsage = errors.New("tls: certificate not authorized for delegation")
+var extensionDelegatedCredential = []int{1, 3, 6, 1, 4, 1, 44363, 44}
+
+// isValidForDelegation returns true if a certificate can be used for delegated
+// credentials.
+func isValidForDelegation(cert *x509.Certificate) bool {
+	// Check that the digitalSignature key usage is set.
+	// The certificate must contains the digitalSignature KeyUsage.
+	if (cert.KeyUsage & x509.KeyUsageDigitalSignature) == 0 {
+		return false
+	}
+
+	// Check that the certificate has the DelegationUsage extension and that
+	// it's non-critical (See Section 4.2 of RFC5280).
+	for _, extension := range cert.Extensions {
+		if extension.Id.Equal(extensionDelegatedCredential) {
+			return true
+		}
+	}
+	return false
+}
+
+// IsExpired returns true if the credential has expired. The end of the validity
+// interval is defined as the delegator certificate's notBefore field ('start')
+// plus ValidTime seconds. This function simply checks that the current time
+// ('now') is before the end of the validity interval.
+func (dc *DelegatedCredential) IsExpired(start, now time.Time) bool {
+	end := start.Add(dc.Cred.ValidTime)
+	return !now.Before(end)
+}
+
+// InvalidTTL returns true if the credential's validity period is longer than the
+// maximum permitted. This is defined by the certificate's notBefore field
+// ('start') plus the ValidTime, minus the current time ('now').
+func (dc *DelegatedCredential) InvalidTTL(start, now time.Time) bool {
+	return dc.Cred.ValidTime > (now.Sub(start) + dcMaxTTL).Round(time.Second)
+}
+
+// Credential stores the public components of a Delegated Credential.
+type Credential struct {
+	// The amount of time for which the credential is valid. Specifically, the
+	// the credential expires 'ValidTime' seconds after the 'notBefore' of the
+	// delegation certificate. The delegator shall not issue delegated
+	// credentials that are valid for more than 7 days from their current time.
+	//
+	// When this data structure is serialized, this value is converted to a
+	// uint32 representing the duration in seconds.
+	ValidTime time.Duration
+	// The signature scheme associated with the credential public key.
+	// This is expected to be the same as the CertificateVerify.algorithm
+	// sent by the server.
+	expCertVerfAlgo SignatureScheme
```

Comment:
Why is this field not exported, but the others are?

Reply:
I unexported the other one, as the only one that needs to be accessed is `PublicKey`